File: 백준/Silver/1874. 스택 수열/스택 수열.cc
#include <iostream>
#include <stack>
using namespace std;

int N;
stack<int> st;

int main(void) {
    string ans = "+\n";
    int cur = 1;
    st.push(cur++);
    cin >> N;
    for (int i = 0; i < N; i++) {
        int tmp;
        cin >> tmp;
        if (st.empty() || tmp > st.top()) {
            do {
                st.push(cur++);
                ans += "+\n";
            } while (tmp > st.top());
            ans += "-\n";
            st.pop();
        } else if (tmp == st.top()) {
            ans += "-\n";
            st.pop();
        } else if (tmp < st.top()) {
            cout << "NO";
            return 0;
        }
    }
    cout << ans;
    return 0;
}

File: 백준/Silver/1874. 스택 수열/스택 수열.py
N = int(input())
st = []
ans = []
cur_num = 1
for _ in range(N):
    e = int(input())
    if st:
        if e > st[-1]:
            while e >= cur_num:
                st.append(cur_num)
                ans.append('+')
                cur_num += 1
            st.pop()
            ans.append('-')
        elif e == st[-1]:
            st.pop()
            ans.append('-')
        elif e < st[-1]:
            ans = "NO"
            break
    else:
        st.append(cur_num)
        ans.append('+')
        cur_num += 1
        while e >= cur_num:
            st.append(cur_num)
            ans.append('+')
            cur_num += 1
        st.pop()
        ans.append('-')
if ans == "NO":
    print(ans)
else:
    print(*ans, sep='\n')
